README.markdown
#1. Name
glypeahead - Port scan through Glype proxies
#2. Author
Junaid Loonat < junaid(at)sensepost(dot)com >
#3. License, version & release date
License : GPL  
Version : v1.1  
Release Date : 2010-04-13
#4. Description
By default, the Glype proxy script has few restrictions on what hosts / ports can be accessed through it. In addition, the proxy script normally displays all cURL-related error messages.  
Using these apparent weaknesses, GlypeAhead is able to perform portscans of targets with reasonable accuracy. Furthermore, the true location of the user is never revealed to the target since all scanning is performed by the remote Glype proxy script.
#5. Requirements
PHP interpreter (with the cURL extension)
#6. Additional Resources 
Blog post - http://www.sensepost.com/blog/4620.html


CHANGELOG
# GlypeAhead Changelog (and other notes)

GlypeAhead 1.1 [2010-04-13]			(Initial Release)

o [Banner Grabbing] If the response starts with a tag (<), then the
  (assumed) webpage is parsed for its TITLE which will then serve as
  the banner for the service. Otherwise, a truncated version of the
  response is used.
  
o [HTTPS Services] Although the port status will be accurately
  determined as either OPEN or CLOSE, it must be noted that banners from
  HTTPS services will vary depending on the Glype configuration,
  targetted HTTPS daemon and / or the port currently being scanned.
  The port however, will still be accurately determined to be either
  - [Glype Configuration]
    Glype is often configured to inform the user that all HTTPS
    requests / responses can still be intercepted.
    This is often when the Glype proxy is running on HTTP, and thus
    all traffic between the client and the proxy will remain unencrypted.
  - [Targetted HTTPS Daemon]
    Upon receiving an HTTP request, many HTTPS daemons will respond
    appropriately by indicating that an HTTPS connection should be
    established instead of HTTP. This behaviour is not always the case.
  - [Port]
    In general, when constructing a request to deliver to the proxy,
    GlypeAhead indicates to the proxy that a HTTP connection should be
    performed / expected.
    When constructing a request to port 443 (a known HTTPS port) however,
    GlypeAhead informs the proxy that a HTTPS connection should be
    performed / expected.
   
o [Error Message Identification] Glype proxies which hide error messages,
  or use custom error messages, cannot be used with GlypeAhead.
  GlypeAhead relies on known / default error messages in order to
  'understand' why the requested page could not be displayed.
  
o [Proxy Cycling] Specified proxies are used in round-robin style for
  each individual portscan.


config.php
<?php
/*
 * GlypeAhead
 * Junaid Loonat (junaid@sensepost.com)
 * Port scanning through glype proxies (which have not been heavily themed / modified)
 * 
 * 13 April 2010
 * 
 * Configuration
 */

$config = array(
	'debug'		=>	false,			//	change to true, of course, to enable debug
	'targets'	=>	array(
		'www.sensepost.com'			=>	array(		//	each targetted port needs to be listed in the array for the target host
			21,
			80,
			443
		),
		'www.hackrack.com'		=>	array(			//	multiple target arrays can be specified
			80,
			8080
		)
	),
	
	/*
	 * Note:
	 * Proxy link should be to Glype's primary index file, index.php
	 * 
	 * Additionally, error messages need to be displayed on resultant page.
	 * This can be checked by appending 'e=curl_error' onto the index url ...
	 * 		http://cooltoday.info/index.php?e=curl_error
	 * ... and looking for the "libcurl returned the error" error message.
	 * 
	 * Currently, the glype proxy cannot be used if it's active theme does not show error messages, or if the proxy's error messages have been customised.
	 */
	'proxies'	=>	array(
		'http://GLYPE.PROXY/index.php',
		'http.//ANOTHER.GLYPE.PROXY/index.php'
	)
);


library/GACurl.php
<?php
/*
 * GlypeAhead
 * Junaid Loonat (junaid@sensepost.com)
 * Port scanning through glype proxies (which have not been heavily themed / modified)
 * 
 * 13 April 2010
 */


/* 
 * GlypeAhead cURL Wrapper
 */
class GACurl
{
	public static function request($opts)
	{
		$ch = curl_init();
		GlypeAhead::displayDebug('cURL request to '.$opts['url']);
		curl_setopt($ch, CURLOPT_URL, $opts['url']);
		curl_setopt($ch, CURLOPT_HEADER, 0);
		curl_setopt($ch, CURLOPT_FOLLOWLOCATION, 1);
		curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
		curl_setopt($ch, CURLOPT_USERAGENT, 'Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 5.1; GlypeAhead 1.1)');
		curl_setopt($ch, CURLOPT_FRESH_CONNECT, 1);
		if (isset($opts['post']))
		{
			curl_setopt($ch, CURLOPT_POST, true);
			curl_setopt($ch, CURLOPT_POSTFIELDS, $opts['post']);
		}
		if (isset($opts['referer']))
		{
			curl_setopt($ch, CURLOPT_REFERER, $opts['referer']);
		}
		$response = array(
			'body'		=>	curl_exec($ch),
			'error'		=>	curl_error($ch),
			'info'		=>	curl_getinfo($ch)
		);
		if (! empty($response['error']))
		{
			GlypeAhead::displayDebug('cURL response generated an error: '.$response['error']);
		}
		GlypeAhead::displayDebug('cURL response HTTP code was '.$response['info']['http_code']);
		curl_close($ch);
		return $response;
	}
}


library/GATarget.php
<?php
/*
 * GlypeAhead
 * Junaid Loonat (junaid@sensepost.com)
 * Port scanning through glype proxies (which have not been heavily themed / modified)
 * 
 * 13 April 2010
 */


/* 
 * GlypeAhead Target Class
 */
class GATarget
{
	private $_host;
	private $_ip;
	private $_ports = array();
	public function load($host, $ports)
	{
		$isValid = false;
		if ($this->addHost($host))
		{
			if (is_array($ports))
			{
				$isValid = true;
				foreach ($ports as $aPort)
				{
					$isValid = $isValid && $this->addPort($aPort);
				}
				if ($isValid)
				{
					rsort($this->_ports);
				}
			} elseif (is_string($ports)) {
				$isValid = $this->addPort($aPort);
			}
		}
		return $isValid;
	}
	private function addHost($host)
	{
		if (is_string($host))
		{
			$host = trim($host);
			$hostIP = gethostbyname($host);
			if (! ip2long($hostIP))
			{
				GlypeAhead::displayError('Failed to resolve target: ' . $host);
			} else {
				$this->_host = $host;
				$this->_ip = $hostIP;
				return true;
			}
		} else {
			GlypeAhead::displayError('Target identifier does not seem to be a valid string');
		}
		return false;
	}
	private function addPort($port)
	{
		$port = trim($port);
		$portN = intval($port);
		if (($portN > 0) && ($portN <= 65535))
		{
			$this->_ports[] = $portN;
			return true;
		} else {
			GlypeAhead::displayError('Invalid port specified for target, ' . $this->_host .': ' . $port);
		}
		return false;
	}
	
	public function unscannedPorts()
	{
		return count($this->_ports);
	}
	public function scanNextPort($proxy)
	{
		$currPortPos = count($this->_ports) - 1;
		$scanResult = array(
			'host'		=>	$this->_host,
			'port'		=>	$this->_ports[$currPortPos]
		);
		$scanResult = array_merge($scanResult, $proxy->getPortInfo($scanResult['host'], $scanResult['port']));
		$scanResult['status'] = $scanResult['status']? 'open' : 'closed';
		array_splice($this->_ports, $currPortPos);
		return $scanResult;
	}
	public function getHost()
	{
		return $this->_host;
	}
	public function getIP()
	{
		return $this->_ip;
	}
	
	
}


library/GlypeAhead.php
<?php
/*
 * GlypeAhead
 * Junaid Loonat (junaid@sensepost.com)
 * Port scanning through glype proxies (which have not been heavily themed / modified)
 * 
 * 13 April 2010
 */


/*
 * GlypeAhead Controller Class
 */
class GlypeAhead
{
	const VERSION = '1.1';
	const RELEASEDATE = '13 April 2010';
	private $_config;
	private $_targets = array();
	private $_proxies = array();
	public static $debugEnabled = false;
	
	public function __construct()
	{
		$this->_config = array();
	}
	public function run($configfile)
	{
		if ($this->loadConfig($configfile))
		{
			if ($this->validateConfig($this->_config))
			{
				if (count($this->_proxies) > 0)
				{
					if (count($this->_targets) > 0)
					{
						$this->doPortScan();
					} else {
						GlypeAhead::displayError('Config-Validation completed but no targets were identified');
					}
				} else {
					GlypeAhead::displayError('Config-Validation completed but no usable proxies were identified');
				}
			} else {
				GlypeAhead::displayError('Issues were detected with the configuration');
			}
		} else {
			GlypeAhead::displayError('Failed to load the specified configuration file');
		}
	}
	private function loadConfig($configfile)
	{
		require_once $configfile;
		if (isset($config) && is_array($config))
		{
			$this->_config = $config;
			return true;
		} elseif (! isset($config)) {
			GlypeAhead::displayError('Failed to locate configuration array from in: ' . $configfile);
		} elseif (! is_array($config)) {
			GlypeAhead::displayError('Invalid configuration array in: ' . $configfile);
		}
		return false;
	}
	private function validateConfig($config)
	{
		$isValid = false;
		if (isset($config['debug']))
		{
			GlypeAhead::$debugEnabled = (bool) $config['debug'];
			if (GlypeAhead::$debugEnabled)
			{
				GlypeAhead::displayDebug('Debug has been enabled');
			}
		}
		foreach (array('targets', 'proxies') as $attrib)
		{
			if (isset($config[$attrib]))
			{
				if (is_array($config[$attrib]))
				{
					foreach ($config[$attrib] as $itemKey=>$itemValue)
					{
						switch ($attrib)
						{
							case 'targets':
								$newTarget = new GATarget();
								if (! $newTarget->load($itemKey, $itemValue))
								{
									GlypeAhead::displayDebug('Dropped target from scan: ' . $itemKey);
									return false;
								} else {
									$this->_targets[] = $newTarget;
								}
								break;
							case 'proxies':
								$newProxy = new GAProxy();
								if (! $newProxy->load($itemValue))
								{
									GlypeAhead::displayDebug('Dropped proxy from scan: ' . $itemValue);
									return false;
								} else {
									$this->_proxies[] = $newProxy;
								}
								break;
							default:
								GlypeAhead::displayError('Uncatered attribute type found: '. $attrib);
								break;
						}
					}
				} else {
					GlypeAhead::displayError('Configuration attribute, '. $attrib .', should be an array');
				}
			} else {
				GlypeAhead::displayError('Missing required configuration attribute: '. $attrib);
			}
		}
		if (count($this->_proxies) >= 1)
		{
			if (count($this->_targets) >= 1)
			{
				$isValid = true;
			} else {
				GlypeAhead::displayError('No targets loaded.');
			}
		} else {
			GlypeAhead::displayError('No proxies loaded.');
		}
		return $isValid;		
	}
	private function doPortScan()
	{
		$currProxy = 0;
		$proxyCount = count($this->_proxies);
		$targetCount = count($this->_targets);
		foreach ($this->_targets as $targetHost)
		{
			GlypeAhead::display('>> Scan report for '.$targetHost->getHost().' ('.$targetHost->getIP().')');
			GlypeAhead::display('   PORT     STATE     BANNER');
			$portCount = 0;
			$scanStart = microtime(true);
			while ($targetHost->unscannedPorts() > 0)
			{
				$scanResult = $targetHost->scanNextPort($this->_proxies[$currProxy++]);
				$currProxy %= $proxyCount;
				GlypeAhead::display('   '.str_pad($scanResult['port'], 5).'    '.str_pad($scanResult['status'], 6).'    '.$scanResult['service']);
				$portCount++;
			}
			$scanEnd = microtime(true);
			$proxyCount = count($this->_proxies);
			GlypeAhead::display('Scanned '.$portCount.' port'.(($portCount > 1)? 's':'').' in '.round(($scanEnd - $scanStart), 2).' seconds, using '.$proxyCount.' prox'.($proxyCount>1?'ies':'y'));
			GlypeAhead::display();
		}
	}
	public static function getHTMLPageTitle($pageBody)
	{
		if (preg_match('#<title>(.+?)</title>#msi', $pageBody, $regs)) {
			return $regs[1];
		}
		return 'Untitled';
	}
	public static function display($msg = '')
	{
		echo $msg . "\n";
	}
	public static function displayError($msg)
	{
		GlypeAhead::display('[!] ' . $msg);
	}
	public static function displayInfo($msg)
	{
		GlypeAhead::display('[i] ' . $msg);
	}
	public static function displayDebug($msg)
	{
		if (GlypeAhead::$debugEnabled)
		{
			GlypeAhead::display('[d] ' . $msg);
		}
	}
	public static function displayHeader()
	{
		GlypeAhead::display('SensePost GlypeAhead v'.GlypeAhead::VERSION.' - released on '.GlypeAhead::RELEASEDATE);
		GlypeAhead::display('Junaid Loonat (junaid@sensepost.com)');
	}
	public static function displayHelp()
	{
		GlypeAhead::display('glypeahead [config-file]');
	}
	public static function fatal($msg)
	{
		GlypeAhead::display('[F] ' . $msg);
		GlypeAhead::display();
		die ();
	}
	public static function stringEndsWith($haystack, $needle)
	{
		$needleLength = strlen($needle);
		if (($needleLength > 0) && (substr_compare($haystack, $needle, -1 * $needleLength, $needleLength) == 0))
		{
			return true;
		}
		return false;
	}
}


library/GAProxy.php
<?php
/*
 * GlypeAhead
 * Junaid Loonat (junaid@sensepost.com)
 * Port scanning through glype proxies (which have not been heavily themed / modified)
 * 
 * 13 April 2010
 */


/* 
 * GlypeAhead Proxy Class
 */
class GAProxy
{
	private $_url;
	
	public function load($url)
	{
		$url = strtolower($url);
		if (GlypeAhead::stringEndsWith($url, 'index.php'))
		{
			$pageResp = GACurl::request(array('url' => $url));
			if ($pageResp['info']['http_code'] == 200)
			{
				$pageResp = GACurl::request(array('url' => $url . '?e='.urlencode('curl_error')));
				if ($pageResp['info']['http_code'] == 200)
				{
					$firstPlaceholder = strpos($pageResp['body'], '%s');
					$lastPlaceholder = strrpos($pageResp['body'], '%s');
					if ($firstPlaceholder != $lastPlaceholder)
					{
						GlypeAhead::displayError('More than one match for cURL error location: ' . $url . '?e=curl_error');
					} else {
						$this->_url = $url;
						if (! $this->isPortOpen('www.sensepost.com'))
						{
							// port is known to be open
							GlypeAhead::displayError('Portscan failed for known open port, www.sensepost.com:80');
						} elseif (! $this->isPortClosed('www.sensepost.com', 1)) {
							// port is known to be closed
							GlypeAhead::displayError('Portscan failed for known closed port, www.sensepost.com:1');
						} else {
							// proxy seems to be working correctly
							return true;
						}
					}
				} else {
					GlypeAhead::displayError('Response for cURL error test failed with a code of ('.$pageResp['info']['http_code'].') on: ' . $url . '?e=curl_error');
				}
			} else {
				GlypeAhead::displayError('Response for direct request to proxy failed with a code of ('.$pageResp['info']['http_code'].') on: ' . $url);
			}
		} else {
			GlypeAhead::displayError('Proxy URL needs to end with "index.php": ' . $url);
		}
		return false;		
	}
	public function isPortClosed($host, $port = 80)
	{
		return ! $this->isPortOpen($host, $port);
	}
	public function isPortOpen($host, $port = 80)
	{
		$pInfo = $this->getPortInfo($host, $port);
		return $pInfo['status'];
	}
	public function getPortInfo($host, $port)
	{
		$pInfo = array(
			'status'	=>	false,
			'service'	=>	''
		);
		$pageResp = $this->request($host, $port);
		$pageRespMsg = $this->parseForProxyMessage($pageResp['body']);
		$pageRespMsg = strtolower($pageRespMsg);
		switch ($pageRespMsg)
		{
			case "couldn't connect to host":
			case "connect() timed out!":
				// port closed or refusing connections
				break;
			case "couldn't resolve host":
				$pInfo['service'] = 'Proxy failed to resolve host';
				break;
			case "empty reply from server":
				$pInfo['status'] = true;
				$pInfo['service'] = '(Empty reply from server)';
				break;
			default:
				$pInfo['status'] = true;
				if (strlen($pInfo['service']) == 0)
				{
					$tmpBody = str_replace(array("\n", "\r"), ' ', $pageResp['body']);
					$tmpBody = trim($tmpBody);
					if ($tmpBody[0] != '<')
					{
						// assume response is not HTML
						$pInfo['service'] = $tmpBody;
					} elseif (strpos($pageResp['body'], 'glype.com') === false) {
						// webpage does not seem to be glype-related
						$pInfo['service'] = '(HTTP) ' . GlypeAhead::getHTMLPageTitle($pageResp['body']);
					} else {
						// may be a glype error page
						$pInfo['status'] = false;
						$pInfo['service'] = '(Glype Error?) ' . $pageRespMsg . ' - ' . GlypeAhead::getHTMLPageTitle($pageResp['body']);
					}
				}
				break;
		}
		if (strlen($pInfo['service']) > 70)
		{
			$pInfo['service'] = substr($pInfo['service'], 0, 70) . ' ...';
		}
		GlypeAhead::displayDebug('Port '.$port.' on host '.$host.' was '.($pInfo['status']?'open':'closed'));
		return $pInfo;
	}
	private function parseForProxyMessage($response)
	{
		$searchStrStart = 'The requested resource could not be loaded. libcurl returned the error:<br><b>';
		$searchStrEnd = '</b>';
		$respString = '';
		$searchStrStartPos = strpos($response, $searchStrStart);
		if ($searchStrStartPos !== false)
		{
			$searchStrStartPos += strlen($searchStrStart);
			$searchStrEndPos = strpos($response, $searchStrEnd, $searchStrStartPos);
			if ($searchStrEndPos !== false)
			{
				$respString = substr($response, $searchStrStartPos, $searchStrEndPos - $searchStrStartPos);
			}
		}
		return trim($respString);
	}
	public function request($host, $port = 80)
	{
		GlypeAhead::displayDebug('Proxied request to '.$host.':'.$port.' through '.$this->_url);
		$targetUrl = '://'.$host.':'.$port;
		if ($port == 443)
		{
			$targetUrl = 's' . $targetUrl;
		}
		$targetUrl = base64_encode($targetUrl);
		$targetUrl = rawurlencode($targetUrl);
		$url = str_replace('index.php', 'browse.php?u='.urlencode($targetUrl).'&b=13&f=', $this->_url);
		$requestOpts = array(
			'url'		=>	$url,
			'referer'	=>	$this->_url
		);
		return GACurl::request($requestOpts);
	}
}
